src/node_http2.cc

```diff
@@ -281,576 +462,526 @@ ssize_t Http2Session::OnCallbackPadding(size_t frameLen,
+      break;
+    case NGHTTP2_PING:
+      session->HandlePingFrame(frame);
+    default:
+      break;
+  }
+  return 0;
 }
 
-void Http2Session::Destroy(const FunctionCallbackInfo<Value>& args) {
-  Http2Session* session;
-  ASSIGN_OR_RETURN_UNWRAP(&session, args.Holder());
-  DEBUG_HTTP2("Http2Session: destroying session %d\n", session->type());
-  Environment* env = Environment::GetCurrent(args);
-  Local<Context> context = env->context();
-
-  bool skipUnconsume = args[0]->BooleanValue(context).ToChecked();
 
-  if (!skipUnconsume)
-    session->Unconsume();
-  session->Close();
+inline int Http2Session::OnFrameNotSent(nghttp2_session* handle,
+                                        const nghttp2_frame* frame,
+                                        int error_code,
+                                        void* user_data) {
+  Http2Session* session = static_cast<Http2Session*>(user_data);
+  Environment* env = session->env();
+  DEBUG_HTTP2SESSION2(session, "frame type %d was not sent, code: %d",
+                      frame->hd.type, error_code);
+  // Do not report if the frame was not sent due to the session closing
+  if (error_code != NGHTTP2_ERR_SESSION_CLOSING &&
+      error_code != NGHTTP2_ERR_STREAM_CLOSED &&
+      error_code != NGHTTP2_ERR_STREAM_CLOSING) {
+    Local<Context> context = env->context();
+    Isolate* isolate = env->isolate();
+    HandleScope scope(isolate);
+    Context::Scope context_scope(context);
+
+    Local<Value> argv[3] = {
+      Integer::New(isolate, frame->hd.stream_id),
+      Integer::New(isolate, frame->hd.type),
+      Integer::New(isolate, error_code)
+    };
+    session->MakeCallback(env->onframeerror_string(), arraysize(argv), argv);
+  }
+  return 0;
 }
 
-void Http2Session::Destroying(const FunctionCallbackInfo<Value>& args) {
-  Http2Session* session;
-  ASSIGN_OR_RETURN_UNWRAP(&session, args.Holder());
-  DEBUG_HTTP2("Http2Session: preparing to destroy session %d\n",
-              session->type());
-  session->MarkDestroying();
+
+inline int Http2Session::OnStreamClose(nghttp2_session* handle,
+                                       int32_t id,
+                                       uint32_t code,
+                                       void* user_data) {
+  Http2Session* session = static_cast<Http2Session*>(user_data);
+  Environment* env = session->env();
+  DEBUG_HTTP2SESSION2(session, "stream %d closed with code: %d", id, code);
+  Http2Stream* stream = session->FindStream(id);
+  // Intentionally ignore the callback if the stream does not exist
+  if (stream != nullptr) {
+    stream->Close(code);
+
+    Isolate* isolate = env->isolate();
+    Local<Context> context = env->context();
+
+    // It is possible for the stream close to occur before the stream is
+    // ever passed on to the javascript side. If that happens, ignore this.
+    if (stream->object()->Has(context,
+                              env->onstreamclose_string()).FromJust()) {
```

Comment:
If this is a likely condition, it might be worth just `Get()`ing the function and passing that to `MakeCallback`.